Firestore/core/src/firebase/firestore/util/to_string.h

```diff
@@ -0,0 +1,202 @@
+std::string ContainerToString(const T& value, std::false_type) {
+  return ObjCToString(value, is_objective_c_pointer<T>{});
+}
+
+template <typename T>
+std::string ContainerToString(const T& value, std::true_type) {
+  std::string contents = absl::StrJoin(
+      value, ", ", [](std::string* out, const typename T::value_type& element) {
+        out->append(ToString(element));
+      });
+  return std::string{"["} + contents + "]";  // NOLINT(whitespace/braces)
+}
+
+// std::string
+
+template <typename T>
+std::string StringToString(const T& value, std::false_type) {
+  return ContainerToString(value, is_iterable<T>{});
+}
+
+template <typename T>
+std::string StringToString(const T& value, std::true_type) {
+  return value;
+}
+
+// Associative container
+
+template <typename T>
+std::string MapToString(const T& value, std::false_type) {
+  return StringToString(value, std::is_convertible<T, std::string>{});
+}
+
+template <typename T>
+std::string MapToString(const T& value, std::true_type) {
+  std::string contents = absl::StrJoin(
+      value, ", ", [](std::string* out, const typename T::value_type& kv) {
+        out->append(
+            StringFormat("%s: %s", ToString(kv.first), ToString(kv.second)));
+      });
+  return std::string{"{"} + contents + "}";  // NOLINT(whitespace/braces)
+}
+
+}  // namespace impl
+
+/**
+ * Creates a human-readable description of the given `value`. The representation
+ * is loosely inspired by Python.
+ *
+ * The general idea is to create the description by using the most specific
+ * available function that creates a string representation of the class; for
+ * containers, do this recursively, adding some minimal container formatting to
+ * the output.
+ *
+ * Example:
+ *
+ * std::vector<DocumentKey> v{
+ *     DocumentKey({"foo/bar"}),
+ *     DocumentKey({"this/that"})
+ * };
+ * assert(ToString(v) == "[foo/bar, this/that]");
+ *
+ * std::map<int, std::string> m{
+       {1, "foo"},
+       {2, "bar"}
+ * };
+ * assert(ToString(m) == "{1: foo, 2: bar}");
+ *
+ * The following algorithm is used:
+ * - if `value` is an associative container (`std::map`, `std::unordered_map`,
+ *   `f:f:immutable::SortedMap`, etc.), the description is of the form:
+ *
+ *     {key1: value1, key2: value2}
+ *
+ *    where the description of each key and value is created by running
+ *    `ToString` recursively;
+ *
+ *  - otherwise, if `value` is (convertible to) `std::string`, it's used as is;
+ *
+ *  - otherwise, if `value` is a container, the description is of the form:
+ *
+ *      [element1, element2]
+ *
+ *    where the description of each element is created by running `ToString`
+ *    recursively;
+ *
+ *  - otherwise, if `value` is an Objective-C class, the description is created
+ *    by calling `[value description]`and converting the result to an
+ *    `std::string`;
+ *
+ *  - otherwise, if `value` defines a member function called `ToString`, the
+ *    description is created by invoking the function;
```

Comment:
If a type is both a (custom) container and defines ToString, then this priority list implies that ToString won't be called. It would seem like ToString() should be given the highest priority rather than the lowest?

Regardless of where you end up on the above question, if you care about this ordering, consider adding a test for it.

Reply:
New ordering seems reasonable. And yeah, the ordering between assoc/simple containers is definitely already tested. The one I was referring to, was the ToString + container ordering, which you've now added. ltgm.